File: 프로그래머스/1/131128. 숫자 짝꿍/숫자 짝꿍.java
import java.util.*;
class Solution {
    public String solution(String X, String Y) {
        String answer = "";
        StringBuilder sb = new StringBuilder();
        int[] x = new int[10];
        int[] y = new int[10];
        
        for(char c : X.toCharArray()) {
            x[c - '0']++;
        }
        for(char c : Y.toCharArray()) {
            y[c - '0']++;
        }
        
        for(int i = 9 ; i >=0; i--) {
            int min = Math.min(x[i],y[i]);
            for(int j = 0; j < min; j++) {
                sb.append(String.valueOf(i));
            }
        }
        answer = sb.toString();
        if(answer.isEmpty()) {
            return "-1";
        }
        else if(answer.charAt(0) == '0') {
            return "0";
        }
        else {
            return answer;
        }
      
        
        
    }
}

File: 프로그래머스/1/131128. 숫자 짝꿍/숫자 짝꿍.js
function createObject(arr,obj) {
     for(let i = 0; i < arr.length;i++) {
        const sNum = arr[i];
        if(!obj[sNum]) obj[sNum] = 1;
        else obj[sNum]++;
    }
}

function solution(X, Y) {
    var answer = '';
    const x = {};
    const y = {};
    createObject(X,x);
    createObject(Y,y);
    const xArr = Object.keys(x);
    const yArr = Object.keys(y);
    
    const newArr = xArr.filter((v) => yArr.includes(v));
    
    for(let i = 0; i < newArr.length; i++) {
        const key = newArr[i];
        const min = Math.min(x[key],y[key]);
        answer += key.repeat(min);
    }
    if(answer.length === 0) return '-1';
    
    answer = answer.split('').sort((a,b)=>b-a).join('');
    if(answer[0] === '0') return '0';
    
    return answer;
}